### PR TITLE
Add helper methods to locate actors that can be reached via a path.

### LINKED ISSUE #17753: Pathfinding performance issues
Can't really vouch for this one being a new thing, but naval AI units put a skirmish game to a halt with just 5 cruisers trying to target my army on the ground out of their reach. Destroying cruisers one by one increased performance considerably, up to VSync-limited 60fps.

https://youtu.be/C9JkW2IwctA

Here's a replay, though it also has a lot of other unneeded stuff. Noted a perfdrop at around 5min, then it got better around 7:20; cruisers chase is closer to the end of the file.
https://www.dropbox.com/s/dh5b2qfgen02ex6/OpenRA-2020-03-01T182317Z.orarep?dl=0

### PR DESCRIPTION
**Pull request by RoosterDragon**
Fixes AI trying to attack units they can't reach, and other automated behaviours will now only consider reachable locations, e.g. if rearming. Fixes #17753

Test scenario:

Start a map with an AI on an island and park some naval units near to their base. Deploy a barracks onto their island but far away from their base and build some units.

On bleed: When the AI creates a squad from the land units it will try and attack the nearest units - the naval units.
This PR: The squad will attack the nearest reachable units - the barracks. The closer naval units are ignored because the land units can't move to them.

----

Previously, the ClosestTo and PositionClosestTo existed to perform a simple distance based check to choose the closest location from a choice of locations to a single other location. For some functions this is sufficient, but for many functions we want to then move between the locations. If the location selected is in fact unreachable (e.g. on another island) then we would not want to consider it.

We now introduce ClosestToIgnoringPath for checks where we don't care about a path existing, e.g. weapons hitting nearby targets. When we do care about paths, we introduce ClosestToWithPathFrom and ClosestToWithPathTo which will check that a path exists. The PathFrom check will make sure one of the actors from the list can make it to the single target location. The PathTo check will make sure the single actor can make it to one of the target locations. This difference allows us to specify which actor will be doing the moving. This is important as a path might exists for one actor, but not another. Consider two islands with a hovercraft on one and a tank on the other. The hovercraft can path to the tank, but the tank cannot path to the hovercraft.

We also introduce WithPathFrom and WithPathTo. These will perform filtering by checking for valid paths, but won't select the closest location.

By employing the new methods that filter for paths, we fix various behaviour that would cause actors to get confused. Imagine an islands map, by checking for paths we ensure logic will locate reachable locations on the island, rather than considering a location on a nearby island that is physically closer but unreachable. This fixes AI squad automation, and other automated behaviours such as rearming.

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/BotModules/Squads/States/GroundStates.cs [hunk 1]
```diff
@@ -90,7 +92,7 @@ public void Tick(Squad owner)
 				}
 			}
 
-			var leader = owner.Units.ClosestTo(owner.TargetActor.CenterPosition);
+			var leader = owner.Units.ClosestToWithPathTo(owner.TargetActor);
```
Comment (dnqbob) [inline OpenRA.Mods.Common/Traits/BotModules/Squads/States/GroundStates.cs]: I think this line may fix part of #17327, by using pathfinding distance to the target, instead of previous geometric distance.

Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/Squads/States/GroundStates.cs]: As for the other part, we need to limit leader's locomotor or use other methods to prevent entire squad stuck on terrain that only leader can pass.

Comment (RoosterDragon) [reply OpenRA.Mods.Common/Traits/BotModules/Squads/States/GroundStates.cs]: This is still using the geometric distance to determine closeness, so it won't aid with that sadly.

Comment (RoosterDragon) [reply OpenRA.Mods.Common/Traits/BotModules/Squads/States/GroundStates.cs]: Have filed #20228 to cover it though.

File: OpenRA.Mods.Common/WorldExtensions.cs [hunk 2]
```diff
@@ -17,6 +17,114 @@ namespace OpenRA.Mods.Common
...
+		/// </summary>
+		public static IEnumerable<Actor> WithPathFrom(this IEnumerable<Actor> actors, Actor sourceActor)
+		{
+			if (sourceActor.TraitOrDefault<Aircraft>() != null)
+				return actors;
+			if (sourceActor.TraitOrDefault<Mobile>() == null)
+				return Enumerable.Empty<Actor>();
+
+			var pathFinder = sourceActor.World.WorldActor.Trait<PathFinder>();
+			var locomotor = sourceActor.Trait<Mobile>().Locomotor;
```
Comment (PunkPun) [inline OpenRA.Mods.Common/WorldExtensions.cs]: we don't need to get `Mobile` twice

File: OpenRA.Mods.Common/WorldExtensions.cs [hunk 3]
```diff
@@ -17,6 +17,114 @@ namespace OpenRA.Mods.Common
...
+		/// </summary>
+		public static WPos? ClosestToWithPathFrom(this IEnumerable<WPos> positions, Actor sourceActor)
+		{
+			if (sourceActor.TraitOrDefault<Aircraft>() != null)
+				return positions.ClosestToIgnoringPath(sourceActor.CenterPosition);
+			if (sourceActor.TraitOrDefault<Mobile>() == null)
+				return null;
+
+			var pathFinder = sourceActor.World.WorldActor.Trait<PathFinder>();
+			var locomotor = sourceActor.Trait<Mobile>().Locomotor;
```
Comment (PunkPun) [inline OpenRA.Mods.Common/WorldExtensions.cs]: same here

**Review by PunkPun — COMMENTED** [review]
Otherwise code LGTM

I've ran a large AI game and noticed some bugs, I'm not sure if they are regressions

Boat can't group
Screenshot 2022-08-28 at 22 44 07

boats try to attack an out of range target
Screenshot 2022-08-28 at 22 47 28

Artilleries unable to attack move
Screenshot 2022-08-28 at 22 48 09

submarines try to move towards visible ground enemy units
Screenshot 2022-08-28 at 22 55 22

**Comment by PunkPun** [comment]
note: in my testing everything went perfectly, I just listed the few exceptions I noticed

File: OpenRA.Mods.Common/WorldExtensions.cs (hunk 2 above)
Comment (RoosterDragon) [reply OpenRA.Mods.Common/WorldExtensions.cs]: Updated

File: OpenRA.Mods.Common/WorldExtensions.cs (hunk 3 above)
Comment (RoosterDragon) [reply OpenRA.Mods.Common/WorldExtensions.cs]: Updated

**Comment by RoosterDragon** [comment]
Hmm - the fact boats will move to attack land at all is actually suspect. That should be prevented by the new checks.

That also reveals a wider issues - we might want things like boats to get close to land to attack things on it, but if the checks were working they wouldn't ever try and do this unless they happened to be in autotarget range. So instead of seeing if we can reach the target, we might want to see if we can reach a location close to the target instead. That would be more complicated.

Will need to investigate this some more.

**Comment by RoosterDragon** [comment]
#20239 fixes things so boats now ignore the land units.

However it seems like a regression to not allow boats to attack units near the shore - we need to teach the AI that it can attack units it can't reach - they don't need to get to the unit - just close enough to be in weapons range. Plus revalidating the other scenarios @PunkPun reported.

**Comment by dnqbob** [comment]
I think it is ok to ignore the land units, for we also have an Autotarget can get target while the squad attack moving

**Review by PunkPun — COMMENTED** [review]
Did another game (rebased on bleed)

The artillery but happened again (ignore the light tank he's attacking the airfield)
Screenshot 2022-09-04 at 13 39 30

Naval grouping seems to be broken. These submarines attacked a naval yard which was built right next to them and did nothing afterwards. When a new sub was built it went kill the enemy naval yard, after that it also went idle. Not that there were anymore targets in water afterwards
Screenshot 2022-09-04 at 13 37 12

In the end, the winning AI was unable to finish the game. There were 4 other AI's which had only the sub pen / naval yard left as their last structure
Screenshot 2022-09-04 at 13 57 56

Probably unrelated to this PR, but AI nuked its own forces. (nuke was launched by the yellow AI)
Screenshot 2022-09-04 at 13 44 41

**Comment by RoosterDragon** [comment]
Pushed a commit that attempts to solve the problem where the AI squad can't directly path to the unit, e.g. boats trying to attack a land unit, but can get close enough to attack, e.g. the land units are on the beach.

Code is godawful and the overall approach still needs some thinking, but this is a step forward. Marking as a draft because more work is needed here.

**Comment by PunkPun** [comment]
I would imagine the proper solution would be to generate a circle around the target the size of weapon range and see if all if any of those cells can be reached.

**Comment by dnqbob** [comment]
I think it could be better that split commit "[Add helper methods to locate actors that can be reached via a path.](https://github.com/OpenRA/OpenRA/pull/20227/commits/3c5103a71fc2cce0c346b8e7fa82f5ab1ebb0f12)" into another PR and we merge that first. AI squad control is a really messing thing that will takes much longer time on fixing and testing.

**Comment by RoosterDragon** [comment]
With only the first commit, AI navy won't attack any land units. This is because they can't path to them. That would probably be worse than the current behaviour of trying to attack land units they can't reach. So we probably do want a solution that finds the balance - navy can attack land units, but only tries to attack land units within range rather than anything.

**Comment by dnqbob** [comment]
> With only the first commit, AI navy won't attack any land units. This is because they can't path to them. That would probably be worse than the current behaviour of trying to attack land units they can't reach. So we probably do want a solution that finds the balance - navy can attack land units, but only tries to attack land units within range rather than anything.

I think it is OK, because "current behaviour of trying to attack land units they can't reach" can give us a huge lag while also stuck the navy on the shore. I personally prefer a smoother gameplay on this issue. I remember RA2 also make AI navy can only attack navy, it is possible that it is due to the same reason.

**Comment by PunkPun** [comment]
It is not fine, see my previous comment

> In the end, the winning AI was unable to finish the game. There were 4 other AI's which had only the sub pen / naval yard left as their last structure

**Comment by dnqbob** [comment]
But I think it is better than when situation happends like https://github.com/MustaphaTR/Romanovs-Vengeance/issues/88, which make the game almost unplayable when there is too many unit trying to attack unreachable.

Add: for optimazing this situation I did many works, like only allow leader do the pathfinding to target, kick stuck unit from squad, but it also happen sometimes when units try to attack unreachable units.

**Comment by anvilvapre** [comment]
Sounds like two different types of path searches, one to reach a destination, another to try to attack something (that will be) in target range?

**Comment by dnqbob** [comment]
Could you please add `BlockedByActor` parameter to those pathfinder helper? To solve the problem of AI stuck by enclosed buildings we need a helper method like that.

**Comment by RoosterDragon** [comment]
We're reliant on `HierarchicalPathFinder.PathExists` for this check. It isn't sufficiently smart enough to help with that I think. It can do `BlockedByActor.None` and some of `BlockedByActor.Immovable` but that's it.

I think something like #20945 is along the right lines anyway. Rather than trying to get clever and working out if placing a building will complete a wall that traps in some units, just enforcing spacing on some/all buildings help reduce the chance of congested AI bases that might trap things in.

**Comment by RoosterDragon** [comment]
I have updated the code to deal with issues noted previously. Here are the scenarios which are now improved.

I used the following map for testing these scenarios out. Human player should use spawn 'A'. Enabling cheats and actor tags are useful. [IslandsPathfinding.zip](https://github.com/OpenRA/OpenRA/files/12087882/IslandsPathfinding.zip)

### Target Units on Same Island

- Check the AI squad target.
- Build a barracks in the bottom-right of your island.
- Build a barracks at the top of the AI island.

On bleed, the squad will initially target your base - the "Assault, fact XX" in the actor tags tells you the squad target. Once the first barracks is placed, they will target that instead - "tent XX". Once the second barracks is placed, they ignore it as the first is physically closer.

With this PR, they won't target the base as it isn't reachable. Same with the first barracks. Once the second barrack is placed, the squad will attack it.


With this PR, the AI squad ignores the closer barracks to move to the one on their island.

### Target Naval Units In Range of Island

- Build a naval yard, then a gunboat.
- Move the gunboat to the AI island shoreline.
- Wait for the AI to react.
- Move the gunboat along the coast, near the shore. Force the AI squad to keep moving to keep up with it.
- Create a second gunboat and place it near the enemy shoreline.
- Move the first gunboat out of range to the left side of the map.

On bleed, the AI squad will attack the gunboat. But when it moves out of range, they won't retarget. They won't respond to new units as a result. With the PR they will drop the target when it moves out of range - and can retarget to the new unit immediately.


Squad targetting the boat.

### Target Lands Units In Range Of Island

- Build some rifles and send them to the stair step at the bottom of your island.
- Wait for the AI squad to react.
- Move the units out of range.

The AI squad will attack these units. On bleed when the units move away the squad won't retarget and will keep trying to attack. With this PR they will go idle and return to base.


The AI squad has dropped the targets that moved out of range.

### AI Units used after squad disbanded

- Set SquadManagerBotModuleInfo.RushInterval to 0 and SquadManagerBotModuleInfo.MinimumAttackForceDelay to 60000. This will ensurce the AI creates a Rush squad.
- Build a Construction Yard on the AI island.
- Wait for the AI squad to destroy it and return to base.
- Build another Construction Yard on the AI island.

On bleed, the squad stays put - they have retargetted to the unreachable construction yard and will be useless. With this PR the units will go idle. Once the next strucuture is placed a new squad is formed once the AI has enough units and will attack your new structure. Importantly, idle units from the first squad will be included in the force.

There existed a bug where an AI squad that disbands doesn't mark those units as available. So the AI never notices they can be assigned into another squad. So the units that returned to base from the first squad would never have been assigned to the new squad without this fix.

----

Finally, I tested a large AI game and didn't notice any performance issues. If you experience any, they are likely to present as lag spikes spaced approx 1 second apart. This is because the AI does some thinking approx every 1 second so it would almost certainly be AI related. If you experience this please include a replay! And if possible some repro steps.

File: OpenRA.Mods.Common/WorldExtensions.cs [hunk 4]
```diff
@@ -11,12 +11,191 @@
...
+		/// <paramref name="sourceActor"/>. Only terrain is taken into account, i.e. as if
+		/// <see cref="BlockedByActor.None"/> was given.
+		/// <paramref name="targetOffsets"/> is used to define locations around each actor in <paramref name="actors"/>
+		/// of which one must be reachable.
+		/// </summary>
+		public static IEnumerable<(Actor Actor, WVec[] ReachableOffsets)> WithPathFrom(this IEnumerable<Actor> actors, Actor sourceActor, Func<Actor, WVec[]> targetOffsets)
+		{
+			if (sourceActor.TraitOrDefault<Aircraft>() != null)
+				return actors.Select<Actor, (Actor Actor, WVec[] ReachableOffsets)>(a => (a, targetOffsets(a)));
+			var mobile = sourceActor.TraitOrDefault<Mobile>();
```
Comment (PunkPun) [inline OpenRA.Mods.Common/WorldExtensions.cs]:
```suggestion
			if (sourceActor.Info.HasTraitInfo<AircraftInfo>())
				return actors.Select<Actor, (Actor Actor, WVec[] ReachableOffsets)>(a => (a, targetOffsets(a)));
			
			var mobile = sourceActor.TraitOrDefault<Mobile>();
```

If we don't use the variable it's better to retrieve the boolean directly

File: OpenRA.Mods.Common/WorldExtensions.cs [hunk 5]
```diff
@@ -11,12 +11,191 @@
...
+		}
+
+		/// <summary>
+		/// Of <paramref name="positions"/> that can be reached as the target of a path from
+		/// <paramref name="sourceActor"/>, returns the nearest by comparing the <see cref="Actor.CenterPosition"/>.
+		/// Only terrain is taken into account, i.e. as if <see cref="BlockedByActor.None"/> was given.
+		/// </summary>
+		public static WPos? ClosestToWithPathFrom(this IEnumerable<WPos> positions, Actor sourceActor)
+		{
+			if (sourceActor.TraitOrDefault<Aircraft>() != null)
```
Comment (PunkPun) [inline OpenRA.Mods.Common/WorldExtensions.cs]: same here

File: OpenRA.Mods.Common/WorldExtensions.cs [hunk 6]
```diff
@@ -11,12 +11,191 @@
...
+		/// <see cref="BlockedByActor.None"/> was given.
+		/// </summary>
+		public static IEnumerable<Actor> WithPathTo(this IEnumerable<Actor> actors, World world, WPos targetPosition)
+		{
+			var pathFinder = world.WorldActor.Trait<PathFinder>();
+			var map = world.Map;
+			return actors
+				.Where(a =>
+				{
+					if (a.TraitOrDefault<Aircraft>() != null)
```
Comment (PunkPun) [inline OpenRA.Mods.Common/WorldExtensions.cs]: and here

File: OpenRA.Mods.Common/WorldExtensions.cs [hunk 7]
```diff
@@ -11,12 +11,191 @@
...
+		/// </summary>
+		public static IEnumerable<(Actor Actor, WPos[] ReachablePositions)> WithPathToAny(
+			this IEnumerable<Actor> actors, World world, Func<Actor, WPos[]> targetPositions)
+		{
+			var pathFinder = world.WorldActor.Trait<PathFinder>();
+			var map = world.Map;
+			return actors
+				.Select<Actor, (Actor Actor, WPos[] ReachablePositions)>(a =>
+				{
+					if (a.TraitOrDefault<Aircraft>() != null)
```
Comment (PunkPun) [inline OpenRA.Mods.Common/WorldExtensions.cs]: and here

File: OpenRA.Mods.Common/Traits/Mobile.cs [hunk 8]
```diff
@@ -925,6 +925,9 @@ void IResolveOrder.ResolveOrder(Actor self, Order order)
 
 			if (order.OrderString == "Move")
 			{
+				if (!order.Target.IsValidFor(self))
+					return;
+
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Traits/Mobile.cs:927]: In a lot of cases these checks aren't necessary as target will be of terrain type making it always true, though I don't oppose the addition of sanity checks. Could they be pushed into their own commit though?

File: OpenRA.Mods.Common/Traits/CaptureManager.cs [hunk 9]
```diff
@@ -221,14 +221,19 @@ public bool StartCapture(Actor self, Actor target, CaptureManager targetManager,
...
 				return false;
 
+			WPos? closestTargetablePosition = null;
+			if (move != null && captures.Info.ConsumedByCapture)
+			{
+				closestTargetablePosition = target.GetTargetablePositions().ClosestToWithPathFrom(self);
+				if (closestTargetablePosition == null)
+					return false;
+			}
+
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Traits/CaptureManager.cs]: why was this code moved outside? Afaik this code shouldn't care about pathfinding, the actor is moving just 1 cell distance into a cell that is likely blocked

File: OpenRA.Mods.Common/Traits/AutoCrusher.cs [hunk 10]
```diff
@@ -63,7 +63,7 @@ void INotifyIdle.TickIdle(Actor self)
 					self.Location != a.Location && a.IsAtGroundLevel() &&
 					Info.TargetRelationships.HasRelationship(self.Owner.RelationshipWith(a.Owner)) &&
 					a.TraitsImplementing<ICrushable>().Any(c => c.CrushableBy(a, self, Info.CrushClasses)))
-				.ClosestTo(self); // TODO: Make it use shortest pathfinding distance instead
+				.ClosestToWithPathFrom(self); // TODO: Make it use shortest pathfinding distance instead
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Traits/AutoCrusher.cs:66]: Comment can be removed?

File: OpenRA.Mods.Common/Activities/Air/ReturnToBase.cs [hunk 11]
```diff
@@ -47,7 +47,7 @@ public static Actor ChooseResupplier(Actor self, bool unreservedOnly)
 					&& a.Owner == self.Owner
 					&& rearmInfo.RearmActors.Contains(a.Info.Name)
 					&& (!unreservedOnly || Reservable.IsAvailableFor(a, self)))
-				.ClosestTo(self);
+				.ClosestToWithPathFrom(self);
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Activities/Air/ReturnToBase.cs:50]: this activity is aircraft exclusive so it doesn't make sense to pathfind

File: OpenRA.Mods.Common/WorldExtensions.cs (hunk 4 above)
Comment (RoosterDragon) [reply OpenRA.Mods.Common/WorldExtensions.cs]: Updated all.

File: OpenRA.Mods.Common/Traits/Mobile.cs (hunk 8 above)
Comment (RoosterDragon) [reply OpenRA.Mods.Common/Traits/Mobile.cs:927]: Made a separate commit.

I had too add this check for AttackMove as the change in the bot code to use that instead of Attack was crashing, so decided to add all missing checks. As you say most of them aren't stricly required but seems better to have them everywhere than to not.

File: OpenRA.Mods.Common/Traits/AutoCrusher.cs (hunk 10 above)
Comment (RoosterDragon) [reply OpenRA.Mods.Common/Traits/AutoCrusher.cs:66]: It's still not the shortest path, so have left it for now. We're doing "closest to as the crow flies" but excluding units where a path doesn't exist. Consider the following contrived scenario:
image

`ClosestToWithPathFrom` will choose A as closest to start, it won't choose b which has a shorter path.

File: OpenRA.Mods.Common/Activities/Air/ReturnToBase.cs (hunk 11 above)
Comment (RoosterDragon) [reply OpenRA.Mods.Common/Activities/Air/ReturnToBase.cs:50]: `ClosestToWithPathFrom` has knowledge that aircraft can path anywhere. So I think I'll keep this as-is so the special casing of Aircraft remains contained within the helpers.

File: OpenRA.Mods.Common/Traits/CaptureManager.cs (hunk 9 above)
Comment (RoosterDragon) [reply OpenRA.Mods.Common/Traits/CaptureManager.cs]: I was wondering what to do if the `closestTargetablePosition` came back as null.

If we can trust it, we don't need this. Reverted the changes and used `ClosestToIgnoringPath` instead.

**Comment by Mailaender** [comment]
Needs a rebase.

File: OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs [hunk 12]
```diff
@@ -194,22 +194,105 @@ void IBotTick.BotTick(IBot bot)
...
+					{
+						targetActor.CenterPosition,
+						targetActor.CenterPosition + new WVec(range, 0, 0),
+						targetActor.CenterPosition + new WVec(-range, 0, 0),
+						targetActor.CenterPosition + new WVec(0, range, 0),
+						targetActor.CenterPosition + new WVec(0, -range, 0),
+						targetActor.CenterPosition + new WVec(rangeDiag, rangeDiag, 0),
+						targetActor.CenterPosition + new WVec(-rangeDiag, rangeDiag, 0),
+						targetActor.CenterPosition + new WVec(-rangeDiag, -rangeDiag, 0),
+						targetActor.CenterPosition + new WVec(rangeDiag, -rangeDiag, 0),
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs:232]: I'm still unsure how to feel about this change. I feel some sort of backup system would still be needed to account for edge cases

Comment (RoosterDragon) [reply OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs:232]: Can you expand on this? Specifically what edge cases do you think are possible and what sort of backup would we need to employ if they occured?

Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs:232]: Edge cases could be something like a river, more chocky areas, or just obstacles miraculously blocking all the 9 cells without actually blocking the target

Comment (RoosterDragon) [reply OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs:232]: If an unfortunate terrain layout means a unit can't be targetted, then yes this will (incorrectly) ignore that unit. If it's the only possible option then indeed we would ignore it in this edge case.

However the following mitigate this scenario in my mind:
- If there's other units, the squad can attack those instead.
- If there's another squad with a different attack range, it might not suffer from the same unfortunate terrain setup and be able to attack.
- If there's another squad in the same domain (e.g. they are on the same island as the target) they can attack the target once available.
- If there's aircraft, they can mop up this target.

The required adverse terrain setup is quite specific that it should be very rare in practise, and if it does occur I think any of the above options mean it shouldn't matter really.

File: OpenRA.Mods.Common/Activities/Hunt.cs [hunk 13]
```diff
@@ -36,7 +36,7 @@ public override bool Tick(Actor self)
 			if (IsCanceling)
 				return true;
 
-			var targetActor = targets.ClosestTo(self);
+			var targetActor = targets.ClosestToWithPathFrom(self);
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Activities/Hunt.cs:39]: hunt will also need distance checks so that  targets on different terrain don't become invalid

Comment (RoosterDragon) [reply OpenRA.Mods.Common/Activities/Hunt.cs:39]: Since it's not caching the target I believe it is always choosing a fresh, valid target. I'm not sure what you'd want me to change here.

**Review by PunkPun — APPROVED** [review]

**Comment by PunkPun** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/059c47dfa1e3a0b3d71461c7e587b2a59461580b...25abdd13a24478592a65685fc6f4dbbf2e496a23)
